### ISSUE TITLE
add concurrency support for eventbus + kotlin + coroutines

### ISSUE THREAD

**Issue by Groostav** [google/guava#3231]
_apologies, I didn't actually mean to post this issue, but since I have, I might as well ask my question, this is really better aimed at stack-overflow though. Again, apologies._

If the `Dispatcher` and `SubscriberFactory`'s were made public, it wouldn't be too difficult to make subscriber methods kotlin-suspend friendly:

```kt
class Publisher {
  fun publish() = launch {
    val eventBus = EventBus()
    val subscribers: List<Subscriber> = generateSubscriberListThroughBusinessLogic()
    for(subscriber in subscribers) {
      eventBus.register(subscriber)
    }

    val publication: Job = eventBus.postAsync("hello!")
    //subscribers would be in process of being invoked

    publication.join()
    //all subscribers are finished
  }
}

class Subscriber(val eventBus: EventBus) {
  @Subscribe suspend fun onMessage(message: String) {
    val value = doDBInteractionOrOtherSuspendingInteraction(message)
  }

  //note the JVM signature of this function is public void onMessage(String message, Continuation $c)
  //so the key in wiring this up is to be able to generate a value for the second parameter at `post()` invocation. 
}
```

But, this would likely require more access to EventBus than it currently offers or seemingly is willing to offer.

Unfortunately this cant be solved easily by using Executors, _at least as far as I know_. I'm willing to write the library, but I would rather _depend on_ guava than fork it.

**Can guava be updated to allow 3rd party support idiomatic kotlin coroutines** without forking guava itself?

**Comment by jbduncan** [comment]
I wonder, might it be worth waiting until [Project](http://openjdk.java.net/projects/loom/) [Loom](http://cr.openjdk.java.net/~rpressler/loom/Loom-Proposal.html) has finished, which I understand is investigating adding coroutines or something similar to Java (on top of its main goal of adding fibers or "green" threads)?

**Comment by Groostav** [comment]
@jbduncan I've spent a great deal of time struggling with this and think I have come to a conclusion. 

Ultimately I forked the event bus, tweaked the `SubscriberRegistry` and added a new dispatch loop that supported kotlin-style _suspension_. I used `Platform.enterNestedEventLoop`, which, I believe, is analagous to what project loom would be doing.

_the result is aweful_: I achieved what I was aiming for, events are dispatched sequentially and long running coroutines will suspend the caller of `post()` --regardless of whether he's java of kotlin. 

But the end result was that performance bugs (IE frame hitching) became _nasty concurrent ones_ instead. UI Element change listeners and buttons would have to be very seriously re-architected such that they acquired necessary locks before calling `eventBus.post()`, which is a very extensive re-architecting.

However, kotlinx-coroutines does have a device to deal with this: when you want to offer concurrency devices downstack without accepting back-pressure from a call, by convention, you make your function take a `CoroutineScope` object, which is the simplest face of kotlin's structured concurrency.

Thus, _I'm pretty sure everything's great_ If i can write:

```kt
@Subscribe fun CoroutineScope.updateWhateverOn(event: SomeEvent){

//jvm signature: 
@Subscribe public void updateWhateverOn(CoroutineScope $receiver, SomeEvent event)
```

Of course, we already have a huge body of regular java subscribers and many kotlin subscribers that do not wish to concern themselves with a `CoroutineScope`, so I would really appreciate it if the event bus provided some facility for a kind of optional parameters/arguments. 

**tl;dr**:
Can support be added for additional parameters to subscribers?

**Comment by Groostav** [comment]
this is related to Kotlin/kotlinx.coroutines#1397

**Comment by jbduncan** [comment]
Sorry @Groostav, I don't have a good response for your query, partly because I'm not that familiar with EventBus and thus I feel out of my depth here, and partly because I'm not actually a Googler and so I don't have the authority to implement your suggestion. :bowing_man: 

@cgdecker Do you have any thoughts regarding this?

**Comment by cgdecker** [comment]
I'm closing this issue because we are no longer going to be making changes to `EventBus` other than important bug fixes. The reasons we're discouraging it and some alternatives are listed in the [EventBus javadoc](https://guava.dev/EventBus).
